File: src/MAPK/models/hatakeyama_2003.py
import equinox as eqx
import jax.numpy as jnp

class hatakeyama_2003(eqx.Module):
    """ Note this is the code at https://models.physiomeproject.org/exposure/c14e77a831ec3f1e56ef03240986a8b7/hatakeyama_kimura_naka_kawasaki_yumoto_ichikawa_kim_saito_saeki_shirouzu_yokoyama_konagaya_2003.cellml/@@cellml_codegen/Python
    
    Which comes from the Hatakeyama paper entry on the CellML site https://models.physiomeproject.org/exposure/c14e77a831ec3f1e56ef03240986a8b7"""
    def __call__(self, t, y, args):
        # unpack state
        R, Shc, PI3K, HRG, R_HRG, R_HRG2, Internalisation, RP, R_Shc, R_ShP, \
            ShP, R_ShGS, ShGS, GS, R_PI3K, R_PI3Kstar, PI3Kstar, RasGTP, \
            RasGDP, Akt_PIPP, RAF_star, RAF, MEKP, MEKPP, P, PIP3, Akt, \
            Akt_PIP3, Akt_PIP, MEK, ERK, ERKP, ERKPP = y
        
        # unpack params
        kf1, kb1, kf2, kb2, kf3, kb3, kf34, kb34, V4, k4, kf5, kb5, kf6, kb6, \
            kf7, kb7, kf8, kb8, kf9, kb9, V10, k10, kf23, kb23, kf24, kb24, \
            kf25, kb25, V26, k26, kf11, k11, V12, k12, kf13, k13, kf14, k14, E, \
            kf27, k27, V28, k28, kf29, kb29, V30, k30, kf31, k31, V32, k32, \
            kf33, k33, k16, k18, PP2A, kf15, k15, kf16, kf17, k17, kf18, MKP3, \
            kf19, k19, kf20, k20, kf21, k21, kf22, k22 = args
        
        v1 = kf1*R*HRG-kb1*R_HRG
        v2 = kf2*R_HRG*R_HRG-kb2*R_HRG2
        v3 = kf3*R_HRG2-kb3*RP
        v4 = (V4*RP)/(k4+RP)
        v5 = kf5*RP*Shc-kb5*R_Shc
        v6 = kf6*R_Shc-kb6*R_ShP
        v7 = kf7*R_ShP*GS-kb7*R_ShGS # old kf7*R_ShP-kb7*R_ShGS
        v8 = kf8*R_ShGS-kb8*ShGS*RP # old kf8*R_ShP-kb8*ShGS*RP
        v9 = kf9*ShGS-kb9*GS*ShP
        v10 = (V10*ShP)/(k10+ShP)
        v11 = (kf11*ShGS*RasGDP)/(k11+RasGDP)
        v12 = (V12*RasGTP)/(k12+RasGTP)
        v13 = (kf13*RasGTP*RAF)/(k13+RAF)
        v14 = (kf14*(Akt_PIPP+E)*RAF_star)/(k14+RAF_star)
        v15 = (kf15*RAF_star*MEK)/(k15*(1.0+MEKP/k17)+MEK)
        v16 = (kf16*PP2A*MEKP)/(k16*(1.0+MEKPP/k18+Akt_PIP/k31+Akt_PIPP/k33)+MEKP)
        v17 = (kf17*RAF_star*MEKP)/(k17*(1.0+MEK/k15)+MEKP)
        v18 = (kf18*PP2A*MEKPP)/(k18*(1.0+MEKP/k16+Akt_PIP/k31+Akt_PIPP/k33)+MEKPP)
        v19 = (kf19*MEKPP*ERK)/(k19*(1.0+ERKP/k21)+ERK)
        v20 = (kf20*MKP3*ERKP)/(k20*(1.0+ERKPP/k22)+ERKP)
        v21 = (kf21*MEKPP*ERKP)/(k21*(1.0+ERK/k19)+ERKP)
        v22 = (kf22*MKP3*ERKPP)/(k22*(1.0+ERKP/k20)+ERKPP)
        v23 = kf23*RP*PI3K-kb23*R_PI3K
        v24 = kf24*R_PI3K-kb24*R_PI3Kstar
        v25 = kf25*R_PI3Kstar-kb25*RP*PI3Kstar
        v26 = (V26*PI3Kstar)/(k26+PI3Kstar)
        v27 = (kf27*PI3Kstar*P)/(k27+P)
        v28 = (V28*PIP3)/(k28+PIP3)
        v29 = kf29*PIP3*Akt-kb29*Akt_PIP3
        v30 = (V30*Akt_PIP3)/(k30*(1.0+Akt_PIP/k32)+Akt_PIP3)
        v31 = (kf31*PP2A*Akt_PIP)/(k31*(1.0+MEKP/k16+MEKPP/k18+Akt_PIPP/k33)+Akt_PIP)
        v32 = (V32*Akt_PIP)/(k32*(1.0+Akt_PIP3/k30)+Akt_PIP)
        v33 = (kf33*PP2A*Akt_PIPP)/(k33*(1.0+MEKP/k16+MEKPP/k18+Akt_PIP/k31)+Akt_PIPP)
        v34 = kf34*RP-kb34*Internalisation


        # define rates
        dy_dt_0 = -v1 # R
        dy_dt_1 = -v5+v10 # Shc
        dy_dt_2 = -v23+v26 # PI3K
        dy_dt_3 = -v1 # HRG
        dy_dt_4 = v1 - 2.0*v2 # R_HRG
        dy_dt_5 = v2-v3+v4 # R_HRG2
        dy_dt_6 = v34 # Internalisation
        dy_dt_7 = v3-v4-v5+v8-v23+v25-v34 # RP
        dy_dt_8 = v5-v6 # R_Shc
        dy_dt_9 = v6-v7 # R_ShP
        dy_dt_10 = v9-v10  # ShP
        dy_dt_11 = v7-v8 # R_ShGS
        dy_dt_12 = v8-v9 # ShGS
        dy_dt_13 = -v7+v9 # GS
        dy_dt_14 = v23-v24 # R_PI3K
        dy_dt_15 = v24-v25 # R_PI3Kstar
        dy_dt_16 = v25-v26 # PI3Kstar
        dy_dt_17 = v11-v12 # RasGTP
        dy_dt_18 = -v11+v12  # RasGDP
        dy_dt_19 = v32-v33 # Akt_PIPP
        dy_dt_20 = -v14+v13 # RAF_star
        dy_dt_21 = v14-v13 # RAF
        dy_dt_22 = v15-v16-v17+v18 # MEKP
        dy_dt_23 = v17-v18 # MEKPP
        dy_dt_24 = v28-v27 # P
        dy_dt_25 = -v28+v27-v29 # PIP3
        dy_dt_26 = -v29 # Akt
        dy_dt_27 = v29-v30+v31 # Akt_PIP3
        dy_dt_28 = v30-v31-v32+v33 # Akt_PIP
        dy_dt_29 = -v15+v16 # MEK
        dy_dt_30 = -v19+v20 # ERK
        dy_dt_31 = v19-v20-v21+v22 # ERKP
        dy_dt_32 = v21-v22 # ERKPP
        
        return (dy_dt_0, dy_dt_1, dy_dt_2, dy_dt_3, dy_dt_4, dy_dt_5, dy_dt_6, \
                    dy_dt_7, dy_dt_8, dy_dt_9, dy_dt_10, dy_dt_11, dy_dt_12, \
                    dy_dt_13, dy_dt_14, dy_dt_15, dy_dt_16, dy_dt_17, dy_dt_18, \
                    dy_dt_19, dy_dt_20, dy_dt_21, dy_dt_22, dy_dt_23, dy_dt_24, \
                    dy_dt_25, dy_dt_26, dy_dt_27, dy_dt_28, dy_dt_29, dy_dt_30, \
                    dy_dt_31, dy_dt_32)

    
    def get_initial_conditions(self):
        """ Function to get nominal initial conditions for the model. """
        R = 80.0
        Shc = 1000.0
        PI3K = 10.0
        HRG = 10.0
        R_HRG = 0.0
        R_HRG2 = 0.0
        Internalisation = 0.0
        RP = 0.0
        R_Shc = 0.0
        R_ShP = 0.0
        ShP = 0.0
        R_ShGS = 0.0
        ShGS = 0.0
        GS = 10.0
        R_PI3K = 0.0
        R_PI3Kstar = 0.0
        PI3Kstar = 0.0
        RasGTP = 0.0
        RasGDP = 120.0
        Akt_PIPP = 0.0
        RAF_star = 0.0
        RAF = 100.0
        MEKP = 0.0
        MEKPP = 0.0
        P = 800.0
        PIP3 = 0.0
        Akt = 10.0
        Akt_PIP3 = 0.0
        Akt_PIP = 0.0
        MEK = 120.0
        ERK = 1000.0
        ERKP = 0.0
        ERKPP = 0.0

        return (R, Shc, PI3K, HRG, R_HRG, R_HRG2, Internalisation, RP, R_Shc, R_ShP, ShP, R_ShGS, ShGS, GS, R_PI3K, R_PI3Kstar, PI3Kstar, RasGTP, RasGDP, Akt_PIPP, RAF_star, RAF, MEKP, MEKPP, P, PIP3, Akt, Akt_PIP3, Akt_PIP, MEK, ERK, ERKP, ERKPP)

        
    
    def get_nominal_params(self):
        """ Function to get nominal parameters for the model. """
        param_dict = {
            'kf1': 0.0012,
            'kb1': 0.00076,
            'kf2': 0.01,
            'kb2': 0.1,
            'kf3': 1.0,
            'kb3': 0.01,
            'kf34': 0.001,
            'kb34': 0.0,
            'V4': 62.5,
            'k4': 50.0,
            'kf5': 0.1,
            'kb5': 1.0,
            'kf6': 20.0,
            'kb6': 5.0,
            'kf7': 60.0,
            'kb7': 546.0,
            'kf8': 2040.0,
            'kb8': 15700.0,
            'kf9': 40.8,
            'kb9': 0.0,
            'V10': 0.0154,
            'k10': 340.0,
            'kf23': 0.1,
            'kb23': 2.0,
            'kf24': 9.85,
            'kb24': 0.0985,
            'kf25': 45.8,
            'kb25': 0.047,
            'V26': 2620.0,
            'k26': 3680.0,
            'kf11': 0.222,
            'k11': 0.181,
            'V12': 0.289,
            'k12': 0.0571,
            'kf13': 1.53,
            'k13': 11.7,
            'kf14': 0.00673,
            'k14': 8.07,
            'E': 7.0,
            'kf27': 16.9,
            'k27': 39.1,
            'V28': 17000.0,
            'k28': 9.02,
            'kf29': 507.0,
            'kb29': 234.0, # TODO check that this reaction is MM
            'V30': 20000.0,
            'k30': 80000.0,
            'kf31': 0.107,
            'k31': 4.35,
            'V32': 20000.0,
            'k32': 80000.0,
            'kf33': 0.211,
            'k33': 12.0,
            'k16': 2200.0,
            'k18': 60.0,
            'PP2A': 11.4,
            'kf15': 3.5,
            'k15': 317.0,
            'kf16': 0.058,
            'kf17': 2.9,
            'k17': 317.0,
            'kf18': 0.058,
            'MKP3': 2.4,
            'kf19': 9.5,
            'k19': 146000.0,
            'kf20': 0.3,
            'k20': 160.0,
            'kf21': 16.0,
            'k21': 146000.0,
            'kf22': 0.27,
            'k22': 60.0,
        }


        param_list = [param_dict[key] for key in param_dict]

        return param_dict, param_list

File: src/MAPK/identifiability/hatakeyama_2003.jl
using StructuralIdentifiability

ode = @ODEmodel(
    x1'(t) = -(kf1*x1(t)*x4(t)-kb1*x5(t)),
    x2'(t) = -(kf5*x8(t)*x2(t)-kb5*x9(t))+((V10*x11(t))/(k10+x11(t))),
    x3'(t) = -(kf23*x8(t)*x3(t)-kb23*x15(t))+((V26*x17(t))/(k26+x17(t))),
    x4'(t) = -(kf1*x1(t)*x4(t)-kb1*x5(t)),
    x5'(t) = (kf1*x1(t)*x4(t)-kb1*x5(t)) - 2.0*(kf2*x5(t)*x5(t)-kb2*x6(t)),
    x6'(t) = (kf2*x5(t)*x5(t)-kb2*x6(t))-(kf3*x6(t)-kb3*x8(t))+((V4*x8(t))/(k4+x8(t))),
    x7'(t) = (kf34*x8(t)-kb34*x7(t)),
    x8'(t) = (kf3*x6(t)-kb3*x8(t))-((V4*x8(t))/(k4+x8(t)))-(kf5*x8(t)*x2(t)-kb5*x9(t))+(kf8*x12(t)-kb8*x13(t)*x8(t))-(kf23*x8(t)*x3(t)-kb23*x15(t))+(kf25*x16(t)-kb25*x8(t)*x17(t))-(kf34*x8(t)-kb34*x7(t)),
    x9'(t) = (kf5*x8(t)*x2(t)-kb5*x9(t))-(kf6*x9(t)-kb6*x10(t)),
    x10'(t) = (kf6*x9(t)-kb6*x10(t))-(kf7*x10(t)*x14(t)-kb7*x12(t)),
    x11'(t) = (kf9*x13(t)-kb9*x14(t)*x11(t))-((V10*x11(t))/(k10+x11(t))) ,
    x12'(t) = (kf7*x10(t)*x14(t)-kb7*x12(t))-(kf8*x12(t)-kb8*x13(t)*x8(t)),
    x13'(t) = (kf8*x12(t)-kb8*x13(t)*x8(t))-(kf9*x13(t)-kb9*x14(t)*x11(t)),
    x14'(t) = -(kf7*x10(t)*x14(t)-kb7*x12(t))+(kf9*x13(t)-kb9*x14(t)*x11(t)),
    x15'(t) = (kf23*x8(t)*x3(t)-kb23*x15(t))-(kf24*x15(t)-kb24*x16(t)),
    x16'(t) = (kf24*x15(t)-kb24*x16(t))-(kf25*x16(t)-kb25*x8(t)*x17(t)),
    x17'(t) = (kf25*x16(t)-kb25*x8(t)*x17(t))-((V26*x17(t))/(k26+x17(t))),
    x18'(t) = ((kf11*x13(t)*x19(t))/(k11+x19(t)))-((V12*x18(t))/(k12+x18(t))),
    x19'(t) = -((kf11*x13(t)*x19(t))/(k11+x19(t)))+((V12*x18(t))/(k12+x18(t))) ,
    x20'(t) = ((V32*x29(t))/(k32*(1.0+x28(t)/k30)+x29(t)))-((kf33*PP2A*x20(t))/(k33*(1.0+x23(t)/k16+x24(t)/k18+x29(t)/k31)+x20(t))),
    x21'(t) = -((kf14*(x20(t)+E)*x21(t))/(k14+x21(t)))+((kf13*x18(t)*x22(t))/(k13+x22(t))),
    x22'(t) = ((kf14*(x20(t)+E)*x21(t))/(k14+x21(t)))-((kf13*x18(t)*x22(t))/(k13+x22(t))),
    x23'(t) = ((kf15*x21(t)*x30(t))/(k15*(1.0+x23(t)/k17)+x30(t)))-((kf16*PP2A*x23(t))/(k16*(1.0+x24(t)/k18+x29(t)/k31+x20(t)/k33)+x23(t)))-((kf17*x21(t)*x23(t))/(k17*(1.0+x30(t)/k15)+x23(t)))+((kf18*PP2A*x24(t))/(k18*(1.0+x23(t)/k16+x29(t)/k31+x20(t)/k33)+x24(t))),
    x24'(t) = ((kf17*x21(t)*x23(t))/(k17*(1.0+x30(t)/k15)+x23(t)))-((kf18*PP2A*x24(t))/(k18*(1.0+x23(t)/k16+x29(t)/k31+x20(t)/k33)+x24(t))),
    x25'(t) = ((V28*x26(t))/(k28+x26(t)))-((kf27*x17(t)*x25(t))/(k27+x25(t))),
    x26'(t) = -((V28*x26(t))/(k28+x26(t)))+((kf27*x17(t)*x25(t))/(k27+x25(t)))-(kf29*x26(t)*x27(t)-kb29*x28(t)),
    x27'(t) = -(kf29*x26(t)*x27(t)-kb29*x28(t)),
    x28'(t) = (kf29*x26(t)*x27(t)-kb29*x28(t))-((V30*x28(t))/(k30*(1.0+x29(t)/k32)+x28(t)))+((kf31*PP2A*x29(t))/(k31*(1.0+x23(t)/k16+x24(t)/k18+x20(t)/k33)+x29(t))),
    x29'(t) = ((V30*x28(t))/(k30*(1.0+x29(t)/k32)+x28(t)))-((kf31*PP2A*x29(t))/(k31*(1.0+x23(t)/k16+x24(t)/k18+x20(t)/k33)+x29(t)))-((V32*x29(t))/(k32*(1.0+x28(t)/k30)+x29(t)))+((kf33*PP2A*x20(t))/(k33*(1.0+x23(t)/k16+x24(t)/k18+x29(t)/k31)+x20(t))),
    x30'(t) = -((kf15*x21(t)*x30(t))/(k15*(1.0+x23(t)/k17)+x30(t)))+((kf16*PP2A*x23(t))/(k16*(1.0+x24(t)/k18+x29(t)/k31+x20(t)/k33)+x23(t))),
    x31'(t) = -((kf19*x24(t)*x31(t))/(k19*(1.0+x32(t)/k21)+x31(t)))+((kf20*MKP3*x32(t))/(k20*(1.0+x33(t)/k22)+x32(t))),
    x32'(t) = ((kf19*x24(t)*x31(t))/(k19*(1.0+x32(t)/k21)+x31(t)))-((kf20*MKP3*x32(t))/(k20*(1.0+x33(t)/k22)+x32(t)))-((kf21*x24(t)*x32(t))/(k21*(1.0+x31(t)/k19)+x32(t)))+((kf22*MKP3*x33(t))/(k22*(1.0+x32(t)/k20)+x33(t))),
    x33'(t) = ((kf21*x24(t)*x32(t))/(k21*(1.0+x31(t)/k19)+x32(t)))-((kf22*MKP3*x33(t))/(k22*(1.0+x32(t)/k20)+x33(t))),
    y1(t) = x33(t)
)

local_id = assess_local_identifiability(ode, 0.99)

# save to file for this model
# create file to save results
file = open("./local_ID_K_2003.txt", "w")
println(file, "Locally Identifiable parameters:")
for (key, value) in local_id
    if value == 1
        print(file, "$key, ")
    end
end
println(file, "")
println(file, "Nonidentifiable parameters:")
for (key, value) in local_id
    if value == 0
        print(file, "$key, ")
    end
end
close(file)

global_id = assess_identifiability(ode, 0.95)

# create file to save results
file = open("./global_ID_K_2003.txt", "w")
println(file, "Globally Identifiable parameters:")
for (key, value) in local_id
    if value == ":globally"
        print(file, "$key, ")
    end
end
println(file, "")
println(file, "Nonidentifiable parameters:")
for (key, value) in local_id
    if value == ":nonidentifiable"
        print(file, "$key, ")
    end
end
close(file)
